feat(filmstrip) Make filmstrip user resizable

robertpin opened the pull request:
Make conference info and toolbar appear on top of the filmstrip
After a breakpoint, filmstrip pushes over the stage view instead of appearing on top
On user resize make tiles wider; after a breakpoint show grid view in the filmstrip
On filmstrip visibility toggle animate stage view resize

robertpin commented:
Jenkins test this please

horymury commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -668,18 +877,21 @@ function _mapStateToProps(state) {
     const { isOpen: shiftRight } = state['features/chat'];
     const {
-        gridDimensions = {},
+        gridDimensions: dimensions = {},
```
why do we need alias here? below you again assign `let gridDimensions = dimensions`. I think you should do a spread of the gridDimensions on another var in case you don't want to alter the original object instead of aliasing, as aliasing does keep the reference.

robertpin replied:
It's because gridDimensions can be reassigned, so it has to be let, not const

horymury commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -306,37 +433,112 @@ class Filmstrip extends PureComponent <Props> {
+        this._dragRef.current.down = true;
+        this._dragRef.current.initialPosition = e.clientX;
+        this._dragRef.current.initialWidth = this.props._verticalFilmstripWidth || DEFAULT_FILMSTRIP_WIDTH;
+        this.setState({
```
nit: extract current into a const  { current } = this._dragRef;

horymury commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -306,37 +433,112 @@ class Filmstrip extends PureComponent <Props> {
+     */
+    _onDragMouseUp() {
+        if (this._dragRef.current.down) {
+            this._dragRef.current.down = false;
```
nit . do we need 2 props for mouseDown bool? We could only keep it in the state prop

horymury commented on react/features/large-video/components/LargeVideo.web.js:
```diff
@@ -126,6 +166,32 @@ class LargeVideo extends Component<Props> {
+            VideoLayout.refreshLayout();
+
+            setTimeout(() => {
+                this._containerRef.current.classList.remove('transition');
```
nit. maybe check if component is still mounted by the time this executes or check this._containerRef/_wrapperRef?.current? , as if it unmounted meanwhile, there might be a crash due to the ref being destroyed.

saghul commented on config.js:
```diff
@@ -1252,6 +1252,10 @@ var config = {
 
+    // Disables user resizable filmstrip. Also, allows configuration of the filmstrip (width, tiles aspect ratios)
+    // through the interfaceConfig options.
+    // disableResizableFilmstrip: false,
```
Can you please create a `filmstrip: { ... }` section instead? This way we can start to migrate those settings in interface_config.js there after this PR lands, instead of having a dozen of scattered around settings.

saghul commented:
Some UI/UX feedback:

<img width="2672" alt="Screenshot 2022-02-02 at 14 53 22" src="https://user-images.githubusercontent.com/317464/152167686-4ccdb704-a2aa-4c31-8ecb-6fd9d2cc7e0e.png">

The filmstrip area is barely noticeable due to the dark on black situation.

The drag handle is very narrow, kinda hard to get a hold of.

The Arrow to close the filmstip is not centered vertically and I cannot unsee that.

<img width="2672" alt="Screenshot 2022-02-02 at 14 53 43" src="https://user-images.githubusercontent.com/317464/152167950-7f13c5f8-5773-454f-b402-6296a0602907.png">

This is what happens when you are alone and start dragging. Maybe we shouldn't allow it?

<img width="1392" alt="Screenshot 2022-02-02 at 14 56 29" src="https://user-images.githubusercontent.com/317464/152168251-31a50f02-7fea-4b0b-88a2-830b512331d5.png">

The gaps around the filmstrip are not consistent.

hristoterezov commented on react/features/base/responsive-ui/actions.js:
```diff
@@ -53,6 +70,7 @@ export function clientResized(clientWidth: number, clientHeight: number) {
                 clientWidth: availableWidth
             });
             dispatch(setAspectRatio(clientWidth, clientHeight));
+            width && dispatch(setFilmstripWidth(width));
```
I'm wondering  if we can move the whole filmstrip logic from here to a middleware in features/filmstrip? WDYT? Maybe it will fit better there?

hristoterezov reviewed (commented):
I'm wondering if we want to implement the same feature for horizontal filmstrip too?

Also I see that the feature is explicitly disabled for mobile browsers. I guess for small screens it doesn't make sense to enable the feature at all but I was wondering if we would like to enable it for tablets?

hristoterezov commented on react/features/base/responsive-ui/actions.js:
```diff
@@ -31,11 +37,14 @@ const REDUCED_UI_THRESHOLD = 300;
 export function clientResized(clientWidth: number, clientHeight: number) {
     return (dispatch: Dispatch<any>, getState: Function) => {
         let availableWidth = clientWidth;
+        let width = null;
```
do we really need `null` here? shouldn't just `let width;` be enough?

hristoterezov reviewed (commented):
Tested to expand the filmstrip on a 2 participant call and I see:
![Screenshot 2022-02-03 at 5 40 30 PM](https://user-images.githubusercontent.com/1729265/152447776-2836af84-99c4-47f3-b055-287b50e20453.png)

Maybe we should kind of change a little bit columns rows calculation for cases like this? WDYT?


Another issue is that a horizontal scrollbar appears from time to time:
![Screenshot 2022-02-03 at 5 50 30 PM](https://user-images.githubusercontent.com/1729265/152448551-b3872eb3-1230-4c8b-9d0c-d83d7da7d9a9.png)
![Screenshot 2022-02-03 at 5 48 13 PM](https://user-images.githubusercontent.com/1729265/152448556-d994576d-3cfb-4a42-b12f-5c93041b5697.png)

hristoterezov commented on react/features/filmstrip/actions.web.js:
```diff
@@ -80,8 +86,41 @@ export function setVerticalViewDimensions() {
-        const thumbnails = calculateThumbnailSizeForVerticalView(clientWidth);
+        const resizableFilmstrip = isFilmstripResizable(state);
+        const thumbnails = resizableFilmstrip
+            ? calculateThumbnailSizeForResizableVerticalView(clientWidth, filmstripWidth.current ?? 0)
```
Why not just put a default value `0` for the `filmstripWidth` in `calculateThumbnailSizeForResizableVerticalView`?

hristoterezov commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -285,17 +394,35 @@ class Filmstrip extends PureComponent <Props> {
             if (!_visible) {
-                filmstripStyle.right = `-${filmstripStyle.maxWidth + 2}px`;
+                if (_resizableFilmstrip) {
+                    filmstripStyle.right = `-${(this._filmstripRef.current?.offsetWidth ?? 0) + 2}px`;
```
Don't we have the filmstrip width in redux? Do we actually need to get it from the reference?

hristoterezov commented on react/features/filmstrip/actions.web.js:
```diff
@@ -80,8 +86,41 @@ export function setVerticalViewDimensions() {
+
+        let gridView = {};
+
+        if (resizableFilmstrip && ((filmstripWidth.current ?? 0) > FILMSTRIP_GRID_BREAKPOINT)) {
```
It looks to me that in this case we don't use calculated thumbnail size from `calculateThumbnailSizeForResizableVerticalView` except the width of the thumbnail in order to calculate the width of the `remoteVideos` div. Is this correct?

If this is the case can we do the whole thing a little bit more obvious what we actually use and calculate?

For example:
I see that the new height value in `dimensions.gridView.filmstripHeight` is actually used for the same element where we use `dimensions.remoteVideosContainer.height`. Maybe we can store the `dimensions.gridView.filmstripHeight` in `dimensions.remoteVideosContainer.height`?
Also in the case where we have grid I don't see why we need to calculate the or pass `local` and `remote` width and height at all. Also in this case the calculation of the `remoteVideosContainer`'s height and width can be done outside `calculateThumbnailSizeForResizableVerticalView`

hristoterezov commented on react/features/filmstrip/functions.web.js:
```diff
@@ -174,6 +177,33 @@ export function calculateThumbnailSizeForVerticalView(clientWidth: number = 0) {
+
+    return {
+        local: {
+            height: DEFAULT_FILMSTRIP_WIDTH,
```
I'm wondering why do we use the `DEFAULT_FILMSTRIP_WIDTH` here and not the ratios for local and remote videos defined in interface_config like `calculateThumbnailSizeForVerticalView`?

robertpin replied:
Because when we resize the filmstrip we want to increase just the width of the tiles, not the height.
![image](https://user-images.githubusercontent.com/11474153/152749930-cbfbabbb-e5cf-4afd-8005-034837439fae.png)

hristoterezov commented on react/features/filmstrip/functions.web.js:
```diff
@@ -174,6 +177,33 @@ export function calculateThumbnailSizeForVerticalView(clientWidth: number = 0) {
+ * @param {number} filmstripWidth - The width of the filmstrip.
+ * @returns {{local: {height, width}, remote: {height, width}}}
+ */
+export function calculateThumbnailSizeForResizableVerticalView(clientWidth: number = 0, filmstripWidth: number) {
```
I don't think there are too many differences between this function and `calculateThumbnailSizeForVerticalView`. Especially if we use the same ratios from interface_config.js instead of `DEFAULT_FILMSTRIP_WIDTH`.

Does it make sense to combine them in one function?

robertpin replied:
Explained [here](https://github.com/jitsi/jitsi-meet/pull/10884#discussion_r800402258) why we don't use the interface config ratios

hristoterezov commented on react/features/filmstrip/functions.web.js:
```diff
@@ -196,13 +227,14 @@ export function calculateThumbnailSizeForTileView({
 
     const minHeight = clientWidth < ASPECT_RATIO_BREAKPOINT ? TILE_MIN_HEIGHT_SMALL : TILE_MIN_HEIGHT_LARGE;
-    const viewWidth = clientWidth - (columns * TILE_HORIZONTAL_MARGIN) - TILE_VIEW_GRID_HORIZONTAL_MARGIN;
+    const viewWidth = clientWidth - (isVerticalFilmstrip ? 0 : columns * TILE_HORIZONTAL_MARGIN)
```
Why do we have an exception for vertical filmstrip?

hristoterezov commented on react/features/filmstrip/functions.web.js:
```diff
@@ -196,13 +227,14 @@ export function calculateThumbnailSizeForTileView({
     const aspectRatioHeight = initialWidth / aspectRatio;
     const noScrollHeight = (clientHeight / rows) - TILE_VERTICAL_MARGIN;
-    const scrollInitialWidth = (viewWidth - SCROLL_SIZE) / columns;
+    const scrollInitialWidth = (viewWidth - (isVerticalFilmstrip ? 0 : SCROLL_SIZE)) / columns;
```
Why do we have an exception for vertical filmstrip?

hristoterezov commented on react/features/filmstrip/reducer.js:
```diff
@@ -92,7 +94,26 @@ const DEFAULT_STATE = {
+         * Current width. Affected by: user filmstrip resize,
+         * window resize, panels open/ close.
+         */
+        current: null,
```
I'm wondering why do you use `null` here and why not `undefined`? 

If you go with `undefined` you can even just do `width: {}`.

robertpin replied:
I could do just `width: {}`, but I wanted to show the object structure and have comments for the properties. And I don't see any benefit of using `undefined` over `null`

hristoterezov replied:
For me the main disadvantage is that null is an object and the typeof checks we may run into problems when doing typeof checks. Also the undefined is the default value for any variable and I just find it easier to maintain in the future. On other hand is there any benefit of using `null` over `undefined` in this case? I'm having a hard time to find any...

Note: I'm not saying you have a bug here and this is just my personal preference.

robertpin replied:
Ok, I understand, I'll change them to undefined, but I'll leave them in with the comments

robertpin replied:
Actually, we need it to be null, because on mount the setVertivalFilmstrip action is dispatched and that converts the initial values into number. Number(null) is 0 which is fine, but Number(undefined) is NaN which is not equal to anything and such keeps triggering the subscriber and causing infinite calls

hristoterezov replied:
Could you please point me to this code?

robertpin replied:
In subscriber.web.js there's a listener for width.current; this is called on mount adn this dispatches setVerticalViewDimensions; this action calls calculateThumbnailSizeForResizableVerticalView and in this function there's a Math.max, which tries to convert to number all the params

hristoterezov commented on react/features/large-video/components/LargeVideo.web.js:
```diff
@@ -126,6 +166,32 @@ class LargeVideo extends Component<Props> {
+
+        if (_resizableFilmstrip && _verticalFilmstripWidth >= FILMSTRIP_BREAKPOINT) {
+            this._containerRef.current.classList.add('transition');
+            this._wrapperRef.current.classList.add('transition');
```
Why don't we do the class addition in the render method? why do we need the reference at all?

robertpin replied:
I tried but it doesn't work. For transitions to work the CSS transition property must be on the element before the modification. If we try to add it on render then the new width and the transition get applied at the same time so the transition is not visible

hristoterezov commented on react/features/large-video/components/LargeVideo.web.js:
```diff
@@ -126,6 +166,32 @@ class LargeVideo extends Component<Props> {
+        if (_resizableFilmstrip && _verticalFilmstripWidth >= FILMSTRIP_BREAKPOINT) {
+            this._containerRef.current.classList.add('transition');
+            this._wrapperRef.current.classList.add('transition');
+            VideoLayout.refreshLayout();
```
Can't we move this to some middleware or something? It will look cleaner since it is legacy code...

robertpin replied:
Not really, because we need to make sure that the transition class gets applied before the recalculation or the transition won't work

hristoterezov commented on react/features/video-layout/functions.js:
```diff
@@ -59,30 +59,32 @@ export function getCurrentLayout(state: Object) {
  * @returns {number}
  */
-export function getMaxColumnCount(state: Object) {
+export function getMaxColumnCount(state: Object, width: ?number = null) {
```
Do we need the default `null` value here? Why not just leave it `undefined`?

hristoterezov commented on react/features/video-layout/functions.js:
```diff
@@ -96,11 +98,12 @@ export function getMaxColumnCount(state: Object) {
  */
-export function getTileViewGridDimensions(state: Object) {
-    const maxColumns = getMaxColumnCount(state);
+export function getTileViewGridDimensions(state: Object, width: ?number = null) {
```
Do we need the default `null` value here? Why not just leave it `undefined`?

hristoterezov commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -243,6 +324,19 @@ class Filmstrip extends PureComponent <Props> {
         super(props);
 
+        this._filmstripRef = React.createRef();
+        this._dragRef = React.createRef();
```
Isn't there more `React`y way of doing the whole drag thing? I mean without references.

hristoterezov commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -243,6 +324,19 @@ class Filmstrip extends PureComponent <Props> {
+        this._dragRef = React.createRef();
+
+        this._dragRef.current = {
+            down: false,
```
Why don't we store those things in the state?

robertpin replied:
I initially used it differently, but yes, they can be stored in state now

hristoterezov commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -306,37 +433,112 @@ class Filmstrip extends PureComponent <Props> {
+    _onDragHandleMouseDown(e) {
+        this._dragRef.current.down = true;
+        this._dragRef.current.initialPosition = e.clientX;
+        this._dragRef.current.initialWidth = this.props._verticalFilmstripWidth || DEFAULT_FILMSTRIP_WIDTH;
```
Is there a reason not to keep all of those in the state?

hristoterezov commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -306,37 +433,112 @@ class Filmstrip extends PureComponent <Props> {
+
+            if (width !== _verticalFilmstripWidth) {
+                dispatch(setUserFilmstripWidth(width));
+                VideoLayout.refreshLayout();
```
Is there a way to move this in a middleware or something since it is legacy code?

hristoterezov reviewed (dismissed):
Left some comments!

In addition switching from the old school filmstrip into tile view filmstrip while dragging was a little shocking. I'm wondering if we can change it in some way?

robertpin commented:
> The filmstrip area is barely noticeable due to the dark on black situation.
Added a different background for filmstrip hover, but on normal (non-hover) it's still gonna look like this.
 
> The drag handle is very narrow, kinda hard to get a hold of.
Fixed.

> The Arrow to close the filmstip is not centered vertically and I cannot unsee that.
Should be fixed now. Could you check again please?

> This is what happens when you are alone and start dragging. Maybe we shouldn't allow it?
WIP.

> The gaps around the filmstrip are not consistent.
I changed the gaps, but they're still not consistent because: on top / bottom we don't want to align them with the title bar/ toolbar according to the design. Also, the right margin is larger because it has space for the scrollbar (old behaviour)

robertpin commented:
> I'm wondering if we want to implement the same feature for horizontal filmstrip too?

No, we don't want to invest any time in the horizontal filmstrip. If anything, we'd like to remove it completely

> Also I see that the feature is explicitly disabled for mobile browsers. I guess for small screens it doesn't make sense to enable the feature at all but I was wondering if we would like to enable it for tablets?

I guess it would make sense on tablets in landscape mode. I'll look into it

robertpin commented:
> Maybe we should kind of change a little bit columns rows calculation for cases like this? WDYT?

I think this should be part of the effort on changing the logic for tile view (like I mentioned on this PR #10860 ). We should change the logic to try to always cover the biggest area.

> Another issue is that a horizontal scrollbar appears from time to time

Should be fixed now. Could you check again please?

robertpin commented:
> In addition switching from the old school filmstrip into tile view filmstrip while dragging was a little shocking. I'm wondering if we can change it in some way?

What do you mean?

hristoterezov commented:
@robertpin I had a quick look, thanks for addressing most of my comments! I still need to retest the PR and I want to do one more detailed pass on the new code!

About the horizontal filmstrip and mobile support - I agree! Thanks for the explanation!

Very important:

> I think this should be part of the effort on changing the logic for tile view (like I mentioned on this PR https://github.com/jitsi/jitsi-meet/pull/10860 ). We should change the logic to try to always cover the biggest area.

So does that mean that we should land the PR after the changes you are mentioning? Or the current state should be fine for now?

And finally:

> In addition switching from the old school filmstrip into tile view filmstrip while dragging was a little shocking. I'm wondering if we can change it in some way?
>What do you mean?

Maybe this will look better when the tile view logic is changed to cover the biggest possible area. But right now when you have 2 people for example you initially see 1 in the bottom and 1 (local) on the top, when you start dragging at some point you see 2 people next to each other at the center of the filmstrip. The transition was a little dramatic for me and I was thinking if we want to smooth it a little bit.

robertpin commented:
@hristoterezov

> So does that mean that we should land the PR after the changes you are mentioning? Or the current state should be fine for now?

I'd like to merge this PR first, as this is done and the for the other one works hasn't even started yet.

> Maybe this will look better when the tile view logic is changed to cover the biggest possible area.

I understand what you're saying, and yes, I think that changing the logic so that we cover the biggest area would help, but as I said, I want to merge this one first.
The logic change is not just for the filmstrip but for the tile view as well, so I think we can do it after

robertpin commented:
Jenkins test this please

robertpin commented:
Jenkins test this please

saghul commented on react/features/filmstrip/actions.web.js:
```diff
@@ -3,17 +3,20 @@ import type { Dispatch } from 'redux';
     SET_VERTICAL_VIEW_DIMENSIONS,
-    SET_VOLUME
+    SET_VOLUME,
+    SET_FILMSTRIP_WIDTH,
```
style: sort alphabetically

saghul commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -20,15 +21,21 @@ import { shouldHideSelfView } from '../../../base/settings/functions.any';
     TILE_VERTICAL_MARGIN,
     TOOLBAR_HEIGHT,
-    TOOLBAR_HEIGHT_MOBILE
+    TOOLBAR_HEIGHT_MOBILE,
```
style: alpha sort.

saghul commented on react/features/filmstrip/actions.web.js:
```diff
@@ -80,21 +85,64 @@ export function setVerticalViewDimensions() {
+        let thumbnails = {};
+        let filmstripDimensions = {};
+
+        if (_verticalViewGrid) {
```
Can you maybe add a one line comment about what specific layout this is "vertical view grid" reads kinda weird to me.

robertpin replied:
Sure. I see your point, but vertical filmstrip is named vertical view in the layouts so I tried to keep it consistent

saghul commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -128,9 +155,9 @@ type Props = {
+     * Whether or not the vertical filmstrip should be displayed as grid.
      */
-    _isToolboxVisible: Boolean,
+    _verticalViewGrid: boolean,
```
alpha sort

saghul commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -161,15 +188,19 @@ const styles = theme => {
             alignItems: 'center',
             justifyContent: 'center',
-            backgroundColor: 'rgba(0, 0, 0, .6)',
+            backgroundColor: 'rgba(51, 51, 51, .5)',
```
This color is repeated several times. Can you pl add a token for it?

robertpin replied:
Discussed it with Tibi and we don't want a token for it as it's just used here. So I just extracted it into a constant

saghul commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -306,37 +459,113 @@ class Filmstrip extends PureComponent <Props> {
+                    this._renderRemoteParticipants()
+                }
+            </div>
+            <AudioTracksContainer />
```
Side note: this should probably not be here. This could be a thing at the same level as dialog container or overlays since it's kinda "global".

robertpin replied:
I'll move it in the parent div, but keep it in this component

saghul commented on react/features/filmstrip/components/web/Thumbnail.js:
```diff
@@ -5,6 +5,7 @@ import clsx from 'clsx';
 import debounce from 'lodash/debounce';
 import React, { Component } from 'react';
 
+import { FILMSTRIP_GRID_BREAKPOINT } from '../..';
```
please import from the constants file directly.

saghul commented on react/features/filmstrip/components/web/Thumbnail.js:
```diff
@@ -32,6 +33,7 @@ import {
     SHOW_TOOLBAR_CONTEXT_MENU_AFTER
 } from '../../constants';
 import { isVideoPlayable, computeDisplayModeFromInput, getDisplayModeInput } from '../../functions';
+import { isFilmstripResizable } from '../../functions.web';
```
Fold this import with the one above?

saghul commented on react/features/filmstrip/constants.js:
```diff
@@ -242,3 +242,31 @@ export const INDICATORS_TOOLTIP_POSITION = {
+ * The minimum width for the stage view
+ * (used to determine the maximum width of the user-resizable vertical filmstrip).
+ */
+export const MIN_STAGE_VIEW_WIDTH = 800;
```
Alpha sort

saghul commented on react/features/filmstrip/functions.web.js:
```diff
@@ -33,7 +34,8 @@ import {
     TILE_MIN_HEIGHT_SMALL,
-    TILE_PORTRAIT_ASPECT_RATIO
+    TILE_PORTRAIT_ASPECT_RATIO,
+    DEFAULT_FILMSTRIP_WIDTH
```
alpha sort

saghul reviewed (commented):
Left some comments, PTAL. I have been unable to test it because I run into a recursion error.

robertpin commented on config.js:
```diff
@@ -1255,6 +1255,13 @@ var config = {
+    // filmstrip: {
+    //     // Disables user resizable filmstrip. Also, allows configuration of the filmstrip
+    //     // (width, tiles aspect ratios) through the interfaceConfig options.
+    //     disableResizableFilmstrip: false,
```
Yes. Or actually disableResizable, so that it's enabled by default

saghul replied:
Yep, whichever better deflects the default indeed.

robertpin commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -489,10 +719,17 @@ class Filmstrip extends PureComponent <Props> {
+
+        if (_verticalViewGrid && (_thumbnailHeight + TILE_VERTICAL_MARGIN) * _rows <= _filmstripHeight) {
+            // additionalClassName = 'vertical-grid-margin';
+            additionalClassName = '';
```
Old code. Oops

robertpin commented on react/features/filmstrip/subscriber.web.js:
```diff
@@ -170,3 +175,12 @@ StateListenerRegistry.register(
+StateListenerRegistry.register(
+    /* selector */ state => state['features/filmstrip'].width?.current,
+    /* listener */(_, store) => {
+        store.dispatch(setVerticalViewDimensions());
```
I don't think so. I mean that value can be changed by either user drag (which is throttled) or by client resize, which isn't throttled, but that seems like it's a different issue. Wdyt?

robertpin commented:
> Should we support resizing when alone? It's weird.

I'm fine either way. 

> There are gaps in the background at the top / bottom, and the thumbnail is not centered.

Fixed the top/bottom ones. The thumbnail is not centered because on the right there is space for the scrollbar. This is not new behavior.

> Does this view make sense? Looks very stretched and doesn't add much, does it? Maybe the threshold can be lowered a bit.

I see your point, but if we lower it we're gonna end up with very small tiles because right now the grid has a min of 2 columns. This will be better to update after Hristo's changes (when one column will be allowed in grid view).


> The arrow doesn't seem to be in the middle.
Should be fixed now.

> With such a small viewport size, it's not possible to drag the filmstrip, and yet the dragging indicator is displayed. I think we should hide it.

I don't know about that. I mean if we hide the indicator when there's not enough width I think it can be kinda weird. I mean opening/ closing a panel (chat or participants) could cause it to hide/ show

hristoterezov commented on react/features/filmstrip/actions.web.js:
```diff
@@ -80,21 +85,65 @@ export function setVerticalViewDimensions() {
+                remoteVideosContainer: _verticalViewGrid ? filmstripDimensions : {
                     width: thumbnails?.local?.width
-                        + TILE_HORIZONTAL_MARGIN + STAGE_VIEW_THUMBNAIL_HORIZONTAL_BORDER + SCROLL_SIZE,
+                        + (TILE_HORIZONTAL_MARGIN / 2) + SCROLL_SIZE,
```
Why do we take 1/2 of `TILE_HORIZONTAL_MARGIN`?

robertpin replied:
In order to reduce the space between the tiles and the scrollbar

hristoterezov commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -668,18 +900,21 @@ function _mapStateToProps(state) {
     const _currentLayout = getCurrentLayout(state);
     const disableSelfView = shouldHideSelfView(state);
+    const _resizableFilmstrip = isFilmstripResizable(state);
+    const _verticalViewGrid = _resizableFilmstrip && verticalFilmstripWidth.current > FILMSTRIP_GRID_BREAKPOINT;
```
I saw the same computations in the actions. Maybe we should move this to a function in order to keep them consistent?

hristoterezov commented on react/features/filmstrip/components/web/Filmstrip.js:
```diff
@@ -202,10 +235,20 @@ const styles = theme => {
 
             '&:hover': {
-                backgroundColor: 'rgba(0, 0, 0, .6)',
+                '& .resizable-filmstrip': {
```
Optional suggestion: I would move the style into separate file since it have grown significantly. WDYT?

robertpin replied:
Good idea

hristoterezov commented on react/features/filmstrip/middleware.web.js:
```diff
@@ -53,6 +57,21 @@ MiddlewareRegistry.register(store => next => action => {
+            }
+
+
+            store.dispatch(setFilmstripWidth(width));
```
Do you think it’s worth it to add a check here and dispatch the action only if we change the value?

I'd imagine that most of the times we'll just dispatch the same value, isn't it?

hristoterezov commented on react/features/filmstrip/functions.web.js:
```diff
@@ -174,6 +177,33 @@ export function calculateThumbnailSizeForVerticalView(clientWidth: number = 0) {
+ * @returns {{local: {height, width}, remote: {height, width}}}
+ */
+export function calculateThumbnailSizeForResizableVerticalView(clientWidth: number = 0, filmstripWidth: number = 0) {
+    const horizontalMargin = VERTICAL_FILMSTRIP_MIN_HORIZONTAL_MARGIN + SCROLL_SIZE
```
I'm not sure if it make sense to merge this function with `calculateThumbnailSizeForVerticalView`?

Anyway I also see the value in having 2 different functions but could you please at least extract the `horizontalMargin` to be a top level constant? It is exactly the same for both functions, isn't it?

robertpin replied:
I wouldn't merge them. I think it's more readable like this. I'll extract the horizontal margin tough

hristoterezov reviewed (commented):
LGTM!

But needs to be rebased!
